### ISSUE TITLE
gRPC server on Python 3.6 is incompatible with os.fork()

### ISSUE THREAD

**Issue by dcoles** [grpc/grpc#17093]
### What version of gRPC and what language are you using?
gRPC 1.16.0 for Python
 
### What operating system (Linux, Windows, …) and version?
 Ubuntu 18.04 LTS (4.15.13-x86_64-linode106) for x86_64
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 Python 3.6.5
 
### What did you do?
1. Checkout the [`python/helloworld/greeter_server.py`](https://github.com/grpc/grpc/blob/master/examples/python/helloworld/greeter_server.py) example
2. Add code that calls fork() to the handler
    - Calling `os.fork()` + `os.execl()`: [greeter_server_fork_exec.py](https://gist.github.com/dcoles/4ed01a8779c800ed841d8b2fc1259e61#file-greeter_server_fork_exec-py-L28)
    - Calling `os.fork()` + `os._exit()`: [greeter_server_fork_exit.py](https://gist.github.com/dcoles/4ed01a8779c800ed841d8b2fc1259e61#file-greeter_server_fork_exit-py-L28)
    - Using `subprocess` + `preexec_fn`: [greeter_server_preexec.py](https://gist.github.com/dcoles/4ed01a8779c800ed841d8b2fc1259e61#file-greeter_server_preexec-py-L32)
3. Start the server with `python3 greeter_server.py`
4. Run `python3 greeter_client.py` multiple times
 
This seems to be the root cause of https://github.com/grpc/grpc/issues/14436.

### What did you expect to see?
```
$ python greeter_client.py
Greeter client received: Hello, you!
$ python greeter_client.py
Greeter client received: Hello, you!
```

### What did you see instead?
**NOTE:** These examples work correctly under Python 2.7 or if using `subprocess` without a `preexec_fn` (does not cause `_posixsubprocess` to call back into Python).

On the server:
```
$ python3.6 greeter_server_preexec.py
E1104 01:27:37.862233836   24328 tcp_server_posix.cc:210]    Failed accept4: Invalid argument
```

On the client:
```
$ python greeter_client.py
Traceback (most recent call last):
  File "greeter_client.py", line 35, in <module>
    run()
  File "greeter_client.py", line 30, in run
    response = stub.SayHello(helloworld_pb2.HelloRequest(name='you'))
  File "/home/dcoles/workspace/grpc/grpc/examples/python/helloworld/venv/local/lib/python3.6/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/home/dcoles/workspace/grpc/grpc/examples/python/helloworld/venv/local/lib/python3.6/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
        status = StatusCode.INTERNAL
        details = "Received RST_STREAM with error code 2"
        debug_error_string = "{"created":"@1541294857.862411425","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Received RST_STREAM with error code 2","grpc_status":13}"
>
$ python greeter_client.py
Traceback (most recent call last):
  File "greeter_client.py", line 35, in <module>
    run()
  File "greeter_client.py", line 30, in run
    response = stub.SayHello(helloworld_pb2.HelloRequest(name='you'))
  File "/home/dcoles/workspace/grpc/grpc/examples/python/helloworld/venv/local/lib/python3.6/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/home/dcoles/workspace/grpc/grpc/examples/python/helloworld/venv/local/lib/python3.6/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
        status = StatusCode.UNAVAILABLE
        details = "Connect Failed"
        debug_error_string = "{"created":"@1541294859.542303123","description":"Failed to create subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":2721,"referenced_errors":[{"created":"@1541294859.542300691","description":"Pick Cancelled","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":241,"referenced_errors":[{"created":"@1541294859.542285256","description":"Connect Failed","file":"src/core/ext/filters/client_channel/subchannel.cc","file_line":689,"grpc_status":14,"referenced_errors":[{"created":"@1541294859.542274056","description":"Failed to connect to remote host: OS Error","errno":111,"file":"src/core/lib/iomgr/tcp_client_posix.cc","file_line":205,"os_error":"Connection refused","syscall":"connect","target_address":"ipv4:127.0.0.1:50051"}]}]}]}"
>
```
 
### Anything else we should know about your project / environment?

gRPC for Python 2.7 and 3.6 was installed using `pip` to ensure I was using the latest version.

**Comment by nicolasnoble** [comment]
How is this usage case different from #16001?

**Comment by srini100** [comment]
@dcoles, this is an interesting problem. How many greeter servers do you see running after sending one request? I see two with python3 and one with python2. I didn't dig into why this is the case but seems like os.execl and os._exit are not really killing the forked child process in python 3.6. Obviously, the child process starts interfering with the parent process when it also tries to accept on the same FD and fails. Child process closes the socket upon error and hence parent process also stops working.

**Comment by dcoles** [comment]
I see the same behaviour. With python2 I start with 1 process containing 4 sub-tasks: 1 thread per request (to a maximum of 10 additional threads). With python3 I also start with 1 process with 4 sub-threads, however when sending the RPC I get 2 new sub-tasks: 1 thread and 1 child process.

I tried debugging the hanging process this via `gdb` and did get a useful backtrace: https://gist.github.com/dcoles/95ca1699de766b5eb2901865c9bf56d8. Looks like it's stuck on `_Server.__del__` waiting for the GRPC server to shutdown due to a bunch of finalizers being called in `PyOS_AfterFork`.

You'll notice that the Python code doesn't even reach `os.execl` or `os._exit` - it's still stuck in `os.fork`. This can be seen by the lack of "Hello, world!" being printed when using `greeter_server_fork_exec.py`.

**Comment by srini100** [comment]
@dcoles, I went as far back as 1.9.0 and see the same problem so this couldn't be due to post fork handlers added in 1.15.0 release to support client side forking. It'll be great if you can investigate this further and find the root cause. There was a mention of this issue about a year ago [here](https://github.com/grpc/grpc/issues/13873#issuecomment-354528889).

**Comment by ericgribkoff** [comment]
Seemingly the same underlying issue: https://github.com/grpc/grpc/issues/15756

**Comment by ericgribkoff** [comment]
I realize that I replied to the PR, but not here on the issue. The full details are at https://github.com/grpc/grpc/pull/17288#issuecomment-443032967, but to summarize: we are doing work in our `__del__` methods to clean up and shut down gRPC that, combined with Python's clean-up of deceased threads in the child's post-fork handler, can lead to a hanging process. We are working to get this fixed in the next release.

**Comment by brandonDu** [comment]
This hava cause some import error, when I use the grpc-python in production with multithread spiders. Sometimes, the grpc client can not connect the grpc server. Thus, the crawl spider failed.

**Comment by ericgribkoff** [comment]
The fix for this (#17444) is now merged and will be available in the next release. The next nightly build at https://packages.grpc.io/ (installable via pip) will also include the fix. I'll close this issue for now.

@dcoles If any issues remain, please comment here and we will reopen. Thanks.
